addons/mail/static/src/model/model_manager.js

```diff
@@ -735,38 +769,31 @@ export class ModelManager {
                     throw Error(`Cannot call _created for already deleted record ${record.localId}.`);
                 }
                 record._created();
-            }
-        }
-        // Execution of "on change".
-        while (this._toCallOnChange.size > 0) {
-            for (const [record, fields] of this._toCallOnChange) {
-                // Delete at every step to detect if the change due to "on change"
-                // registered extra fields for which to call "on change".
-                this._toCallOnChange.delete(record);
-                if (!record.exists()) {
-                    throw Error(`Cannot execute 'on change' for already deleted record ${record.localId}.`);
-                }
-                while (fields.size > 0) {
-                    for (const field of fields) {
-                        // Delete at every step to detect if the change due to "on change"
-                        // registered extra fields for which to call "on change".
-                        fields.delete(field);
-                        if (field.compute) {
-                            const res = record[field.compute]();
-                            if (res !== undefined) {
-                                throw new Error("'on change' compute method is not supposed to return any value.");
-                            }
-                            continue;
+                for (const [dependencies, onChange] of record.constructor.onChanges || []) {
+                    const listener = {
+                        onChange() {
+                             record[onChange]();
+                        },
```

Comment:
Ouch... It's too hacky to put technical stuffs in static class field `onChanges` dynamically and have function name store for evaluation here.

The code becomes very hard to understand and maintain when storing dynamic stuffs on items themselves that are not inherent to the items themselves.

A better solution is to properly define an attribute of model manager, which properly document what's in there.

Reply:
Ehm... I may have commented this before actually see how it's been defined.

Actually it's not dynamically added, so it's fine on this aspect.
The way to define this sucks, but it's more tolerable than I though.

Reply:
I guess I can also create a class for defining their structure if necessary?

Reply:
> I guess I can also create a class for defining their structure if necessary?

Good idea! It would be nicer to use, and the code of model manager will feel less hacky that way

Reply:
BTW, just to point out that classes are kind-of low-cost models when used that way. If you keep them as plain object generators without built-in behaviour (i.e. "methods") and no inheritance, then they are decent abstraction level.

They lack many features of true models, so use only them in core model code and outside we should use model all the time.

_(Also these classes will be easy to convert to model in the future, as I plan model core to work internally with models and fields in the future... Basically bootstrapping. This is will improve drastically the maintainability of model core code because all the flows and registries will be plain **computed** fields)_